### PR TITLE
Add Watch Conformance Tests

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This PR adds the conformance test runners. The tests are run as part of the integration test suite.

### CODE REVIEW

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/ActionFilter.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+import com.google.common.base.Splitter;
+import com.google.firestore.v1.Document;
+import com.google.firestore.v1.RunQueryRequest.ConsistencySelectorCase;
+import com.google.firestore.v1.StructuredQuery;
+import com.google.firestore.v1.Value;
+import java.util.List;
+import java.util.function.Predicate;
+
+/** Filters test traces for conformance testing. */
+public final class ActionFilter implements Predicate<DatastoreTestTrace.DatastoreAction> {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/ActionFilter.java:31]: java/com/google/firebase/datastore/clients/testing/conformance/preprocessor/ActionFilter.java

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/ConformanceRuntime.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * A runtime for performing Firestore conformance tests on Android.
+ *
+ * <p>This class contains all logic for a test run. Test cases only need to supply data (and this
+ * should be done by the test code generator).
+ */
+public final class ConformanceRuntime {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/ConformanceRuntime.java:45]: java/com/google/firebase/datastore/clients/testing/conformance/runtime/ConformanceRuntime.java

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/TestCaseConverter.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+import com.google.firestore.v1.Document;
+import com.google.firestore.v1.RunQueryResponse;
+import com.google.firestore.v1.StructuredQuery;
+import com.google.firestore.v1.Value;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class TestCaseConverter {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/TestCaseConverter.java:37]: java/com/google/firebase/datastore/clients/testing/conformance/preprocessor/TraceConverter.java

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/TestCaseIgnoreList.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+/**
+ * Filters testcases by name.
+ *
+ * <p>The ignore list contains regular expressions matching testcase names. There should be one
+ * expression per line. Comments are supported at the beginning or end of lines using the {@code #}
+ * symbol. Empty lines are ignored.
+ */
+public final class TestCaseIgnoreList implements Predicate<TestCase> {
+  // Note: This code is copied from Google3.
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/TestCaseIgnoreList.java:35]: java/com/google/firebase/datastore/clients/testing/conformance/preprocessor/TestcaseBlacklist.java

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/TestCollection.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+// limitations under the License.
+
+package com.google.firebase.firestore.conformance;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** A simplified representation of a Firestore collection for testing. */
+public final class TestCollection {
+  // Note: This code is copied from Google3.
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/TestCollection.java:22]: java/com/google/firebase/datastore/clients/testing/conformance/runtime/TestCollection.java

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/TestDocument.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.firestore.conformance;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/** A simplified representation of a Firestore document for testing. */
+public final class TestDocument {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/TestDocument.java:21]: java/com/google/firebase/datastore/clients/testing/conformance/runtime/TestDocument.java

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Collection.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+
+import androidx.annotation.NonNull;
+import com.google.auto.value.AutoValue;
+import com.google.firestore.v1.Document;
+import java.util.Collections;
+import java.util.List;
+
+/** A collection of documents. */
+@AutoValue
+public abstract class Collection {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Collection.java:25]: firebase/datastore/clients/testing/conformance/test.proto

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Mutation.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+// limitations under the License.
+
+package com.google.firebase.firestore.conformance.model;
+
+import androidx.annotation.NonNull;
+import com.google.auto.value.AutoValue;
+import com.google.firestore.v1.Document;
+
+@AutoValue
+public abstract class Mutation {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Mutation.java:22]: firebase/datastore/clients/testing/conformance/test.proto

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Order.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+package com.google.firebase.firestore.conformance.model;
+
+import androidx.annotation.NonNull;
+import com.google.auto.value.AutoValue;
+import com.google.firestore.v1.StructuredQuery;
+
+/** An order on a field. */
+@AutoValue
+public abstract class Order {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Order.java:23]: firebase/datastore/clients/testing/conformance/test.proto

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Query.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+package com.google.firebase.firestore.conformance.model;
+
+import androidx.annotation.NonNull;
+import com.google.auto.value.AutoValue;
+import java.util.Collections;
+import java.util.List;
+
+/** A query call stack to Firestore. */
+@AutoValue
+public abstract class Query {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Query.java:24]: firebase/datastore/clients/testing/conformance/test.proto

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/QueryFilter.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+
+package com.google.firebase.firestore.conformance.model;
+
+import androidx.annotation.NonNull;
+import com.google.auto.value.AutoValue;
+import com.google.firestore.v1.Value;
+import javax.annotation.Nullable;
+
+@AutoValue
+public abstract class QueryFilter {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/QueryFilter.java:23]: firebase/datastore/clients/testing/conformance/test.proto

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Result.java [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+
+import androidx.annotation.NonNull;
+import com.google.auto.value.AutoValue;
+import com.google.firestore.v1.Document;
+import java.util.Collections;
+import java.util.List;
+
+/** Result from query execution. */
+@AutoValue
+public abstract class Result {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Result.java:25]: firebase/datastore/clients/testing/conformance/test.proto

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/TestCase.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+
+package com.google.firebase.firestore.conformance.model;
+
+import androidx.annotation.NonNull;
+import com.google.auto.value.AutoValue;
+import java.util.List;
+import javax.annotation.Nullable;
+
+@AutoValue
+public abstract class TestCase {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/TestCase.java:23]: firebase/datastore/clients/testing/conformance/test.proto

File: firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Where.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+package com.google.firebase.firestore.conformance.model;
+
+import androidx.annotation.NonNull;
+import com.google.auto.value.AutoValue;
+import com.google.firestore.v1.StructuredQuery;
+import com.google.firestore.v1.Value;
+
+/** A Firestore query filter. */
+@AutoValue
+public abstract class Where {
```
Comment (schmidt-sebastian) [inline firebase-firestore/src/androidTest/java/com/google/firebase/firestore/conformance/model/Where.java:24]: firebase/datastore/clients/testing/conformance/test.proto

File: firebase-firestore/src/proto/google/firebase/firestore/proto/bundle.proto [hunk 15]
```diff
@@ -11,7 +11,6 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-//
```
Comment (thebrianchen) [inline firebase-firestore/src/proto/google/firebase/firestore/proto/bundle.proto:14]: Nice attention to detail :P

**Review by thebrianchen — APPROVED** [review]
lgtm

**Pull request merged**